sphinx/locale/__init__.py

```diff
@@ -112,9 +112,21 @@ def init(
     # the None entry is the system's default locale path
     has_translation = True
 
-    if language and '_' in language:
+    if getenv('SOURCE_DATE_EPOCH') is not None:
+        # Disable localization during reproducible source builds
+        # See https://reproducible-builds.org/docs/source-date-epoch/
+        #
+        # Note: Providing an empty/none value to gettext.translation causes
+        # it to consult various language-related environment variables to find
+        # locale(s).  We don't want that during a reproducible build; we want
+        # to run through the same code path, but to return NullTranslations.
+        #
+        # To achieve that, specify the ISO-639-3 'undetermined' language code,
+        # which should not match any translation catalogs.
+        languages: Optional[List[str]] = ['und']
```

Comment:
This may seem overly complicated compared to, for example:

```python
if getenv('SOURCE_DATE_EPOCH') is not None:
    return NullTranslations(), False
```

...which would achieve similar results.

My thinking here is that the presence/absence of `SOURCE_DATE_EPOCH` should, ideally, minimize the effect on the control flow of the program and build.

If the environment variable results in completely different control flows, then arguably the program being built _is_ different.

My sense is that the ideal roadmap for `SOURCE_DATE_EPOCH` is to introduce it as-required to achieve reproducible builds, but also to keep in mind that it should be removed in future when no-longer-required.

(and to be precise in the case of `sphinx`: a potential time for removal would be when it's possible to perform deterministic documentation builds in all-available-locales -- meaning that the software would be packaged for universal applicability and should build into an identical artefact regardless of the build host's locale)

Reply:
> (and to be precise in the case of `sphinx`: a potential time for removal would be when it's possible to perform deterministic documentation builds in all-available-locales -- meaning that the software would be packaged for universal applicability and should build into an identical artefact regardless of the build host's locale)

If it is realistic to achive this, I think it's a good goal -- would you be happy to open a new issue to track the 'locale independent' workstream?

A

Reply:
From searching around before doing that, I think that this would be approximately the same as #788 (although that is specific to HTML, currently).

I'll spend some more time reading before deciding whether to add more comments there or to open a separate ticket.

Reply:
Thank you, I appreciate your help here!

A